service/matching/db.go

```diff
@@ -165,26 +167,44 @@ func (db *taskQueueDB) renewTaskQueueLocked(
 	return nil
 }
 
-// UpdateState updates the queue state with the given value
-func (db *taskQueueDB) UpdateState(
+// UpdateStateAckLevelBacklogCount updates the queue state with the given Ack and BacklogCount values
+func (db *taskQueueDB) UpdateStateAckLevelBacklogCount(
 	ctx context.Context,
 	ackLevel int64,
+	approximateBacklogCount int64,
 ) error {
 	db.Lock()
 	defer db.Unlock()
 	queueInfo := db.cachedQueueInfo()
 	queueInfo.AckLevel = ackLevel
+	queueInfo.ApproximateBacklogCount = approximateBacklogCount
 	_, err := db.store.UpdateTaskQueue(ctx, &persistence.UpdateTaskQueueRequest{
 		RangeID:       db.rangeID,
 		TaskQueueInfo: queueInfo,
 		PrevRangeID:   db.rangeID,
 	})
 	if err == nil {
 		db.ackLevel = ackLevel
+		db.approximateBacklogCount = approximateBacklogCount
 	}
 	return err
 }
 
+// updateInMemoryBacklogCount updates the in-memory DB state with the given approximateBacklogCount value
+func (db *taskQueueDB) updateInMemoryBacklogCount(
+	approximateBacklogCount int64,
+) {
+	db.Lock()
+	defer db.Unlock()
+	db.approximateBacklogCount += approximateBacklogCount
+}
+
+func (db *taskQueueDB) getApproximateBacklogCount() int64 {
+	db.Lock()
+	defer db.Unlock()
```

Comment:
I was thinking of implementing a similar approach to what @dnr had recommended [here]. Change `approximateBacklogCounter` to be an atomic and inside of `getApproximateBacklogCount`, simply load the atomic and write to it atomically while holding the db lock. Think this is useful since taskReader also calls `getApproximateBacklogCount` and we don't want it to wait until db's lock has been held. @ShahabT (https://github.com/temporalio/temporal/pull/5675#discussion_r1556496491)

Reply:
This has been implemented in the latest push.